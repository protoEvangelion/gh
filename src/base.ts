// https://oclif.io/docs/base_class.html

import * as simpleGit from 'simple-git/promise'
import { find } from 'lodash'
import Command, { flags } from '@oclif/command'
import { log } from './logger'
import { config } from './config'

const git = simpleGit()

export default abstract class extends Command {
  public static flags: any = {
    help: flags.help({ char: 'h' }),
    debug: flags.boolean({
      description: 'A more complete info flag, which leaks more privacy sensitive data by default.',
    }),
    info: flags.boolean({
      description: 'The info flag is useful for basic debugging',
    }),
    remote: flags.string({
      description: 'Override the default_remote setting in ~/.default.gh.json',
    }),
    repo: flags.string({ char: 'r', description: 'The repo to fetch issues from' }),
    user: flags.string({ char: 'u', description: 'The owner of the repository' }),
  }

  public flags
  public remoteInfo

  public async init() {
    try {
      var isGitRepo = await git.checkIsRepo()
    } catch (e) {
      throw new Error(`Error when checking if current dir is a git repository: ${e}`)
    }

    if (!isGitRepo) {
      throw new Error('Current directory is not a git repo')
    }

    this.setGlobalFlags()
    this.initLogger()

    await this.setUserAndRepo()
  }

  private initLogger() {
    process.env.DEBUG = this.flags.debug
    process.env.INFO = this.flags.info
  }

  private setGlobalFlags() {
    // @ts-ignore: need to figure out if this error is benign
    const { flags } = this.parse(this.constructor)

    this.flags = flags
  }

  private async setUserAndRepo() {
    // TODO: add default remote to config https://github.com/node-gh/gh/issues/85
    try {
      var remotesArr = await git.getRemotes(true)
    } catch (e) {
      throw new Error(`Error when looking up your local git remotes: ${e}`)
    }

    const remoteName = this.flags.remote || config.default_remote || 'origin'

    try {
      // @ts-ignore: https://github.com/steveukx/git-js/pull/283
      var remote = find(remotesArr, { name: remoteName }).refs.fetch
    } catch (e) {
      throw new Error(
        `No remote with name: ${remoteName}. Try adding one with \`git remote add origin your_repo_fetch_url\``
      )
    }

    // @ts-ignore: https://github.com/steveukx/git-js/pull/283
    if (remotesArr.length === 1) {
      remote = remotesArr[0].refs.fetch
    }
    const user = this.flags.user ? this.flags.user : remote.match('github[.]com.(.*)/')![1]
    const repo = this.flags.repo ? this.flags.repo : remote.match('/(.*)[.]git')![1]

    this.remoteInfo = {
      remote,
      repo,
      user,
    }

    log.debug(`Remote: ${remoteName} \n`, `Repo: ${repo} \n`, `Owner: ${user} \n`)
  }

  /*
  * This method exists for backwards compatibility of plugins for v1.x
  */
  public static getConfig() {
    return config
  }
}
