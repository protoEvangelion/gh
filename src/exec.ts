/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as childProcess from 'child_process'
import * as logger from './logger'

export function spawnSync(cmd, args, options?: object) {
    let exec

    logger.debug(`spawnSync: ${cmd} ${args.join(' ')}`)

    exec = childProcess.spawnSync(cmd, args, options)

    if (exec.error && exec.error.code === 'ENOENT' && process.platform === 'win32') {
        logger.debug("Invoking patched sapwnSync due to Windows' libuv bug")
        exec = childProcess.spawnSync(`${cmd}.cmd`, args, options)
    }

    return {
        stdout: exec.stdout.toString().trim(),
        stderr: exec.stderr.toString().trim(),
        status: exec.status,
    }
}

export function spawnSyncStream(cmd, args, options?: any) {
    let proc
    let err

    if (!options) {
        options = {}
    }

    options.stdio = ['pipe', process.stdout, process.stderr]

    logger.debug(`spawnSyncStream: ${cmd} ${args.join(' ')}`)

    proc = childProcess.spawnSync(cmd, args, options)

    if (proc.status !== 0) {
        err = new Error()
        err.code = proc.status
        err.message = `Child process terminated with error code ${err.code}`

        throw err
    }

    return proc
}

export function execSync(cmd, options?) {
    if (!options) {
        options = {}
    }

    logger.debug(`execSync: ${cmd}`)

    options.stdio = ['pipe', process.stdout, process.stderr]

    return childProcess.execSync(cmd, options)
}

export function execSyncInteractiveStream(cmd, options?) {
    if (!options) {
        options = {}
    }

    logger.debug(`execSyncInteractiveStream: ${cmd}`)

    options.stdio = 'inherit'

    return childProcess.execSync(cmd, options)
}
