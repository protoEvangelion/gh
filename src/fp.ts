/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as S from 'sanctuary'
import * as Future from 'fluture'
import * as fs from 'fs'
import * as which from 'which'
import * as $ from 'sanctuary-def'

/* SAFE UTILS */
export const safeWhich = Future.encase(which.sync)
export const safeRealpath = Future.encase(fs.realpathSync)

export function safeImport(
    fileName: string
): Future.FutureInstance<NodeJS.ErrnoException, object | string> {
    let moduleObj = null

    try {
        moduleObj = require(fileName)
    } catch (e) {
        return e.code === 'MODULE_NOT_FOUND' ? Future.of(fileName) : Future.reject(e)
    }

    return Future.of(moduleObj)
}

export const safeReaddir = Future.encaseN<any, string[], string>(fs.readdir)

/* TRANSFORMATIONS */
// Allows you to take a Maybe returning function and make it an Either returning function
export function maybeFnToEither(monadReturningFunction) {
    return function convertMaybe(val) {
        const result = monadReturningFunction(val)

        return S.isJust(result) ? S.Right(result.value) : S.Left(null)
    }
}

/**
 * Concats two strings
 * @return {Future}
 */
export const prepend = (a: string) => (b: string) => {
    const argsAreStrings = typeof a === 'string' && typeof b === 'string'

    return argsAreStrings ? Future.of(a + b) : Future.reject('Both args should be strings')
}

/* TYPE CHECKING UTILS */
export const isObject = S.is($.Object)
export const isString = S.is($.String)

/* LOGGING UTILS */
export const l = x => {
    console.log('!!!!!!!', x)
    return x
}

// returns a future
export const lf = x => {
    console.log('-------', x)
    return Future.of(x)
}
