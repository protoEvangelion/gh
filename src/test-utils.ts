/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import { execSync } from 'child_process'
import { isArray, isObject, isPlainObject, map, mapValues, upperFirst } from 'lodash'
import * as nock from 'nock'
import * as zlib from 'zlib'

export function runCmd(cmd: string, env?: boolean) {
    try {
        const customEnv = env ? { env: { ...process.env } } : {}

        var result = execSync(cmd, { cwd: process.cwd(), ...customEnv })
    } catch (error) {
        throw new Error(error.output.toString())
    }

    return result.toString()
}

const nockBack = nock.back

export function prepareTestFixtures(cmdName, argv) {
    let id = 0

    // These should only include the flags that you need for e2e tests
    const cmds = [
        {
            name: 'Help',
            flags: [],
        },
        {
            name: 'Issue',
            flags: ['--comment', '--new', '--open', '--close', '--search', '--assign'],
        },
        {
            name: 'PullRequest',
            flags: [
                '--detailed',
                '--info',
                '--fetch',
                '--fwd',
                '--comment',
                '--open',
                '--close',
                '--draft',
                '--submit',
            ],
        },
        {
            name: 'Gists',
            flags: ['--new', '--fork', '--delete'],
        },
        {
            name: 'Milestone',
            flags: ['--list'],
        },
        {
            name: 'Notifications',
        },
        {
            name: 'Repo',
            flags: ['--label', '--list', '--new', '--fork', '--delete', '--search'],
        },
        {
            name: 'User',
            flags: ['--login', '--logout', '--whoami'],
        },
        {
            name: 'Version',
            flags: ['--version'],
        },
    ].filter(cmd => cmd.name === cmdName)

    const newCmdName = formatCmdName(cmds[0], argv)

    if (!newCmdName) {
        return () => {}
    }

    nock.disableNetConnect()
    nockBack.fixtures = `${process.cwd()}/__tests__/nockFixtures`
    nockBack.setMode('record')

    const nockPromise = nockBack(`${newCmdName}.json`, {
        before,
        afterRecord,
    })

    return () =>
        nockPromise
            .then(({ nockDone }) => nockDone())
            .catch(err => {
                throw new Error(`Nock ==> ${err}`)
            })

    /* --- Normalization Functions --- */

    function normalize(value, key) {
        if (!value) return value

        if (isPlainObject(value)) {
            return mapValues(value, normalize)
        }

        if (isArray(value) && isPlainObject(value[0])) {
            return map(value, normalize)
        }

        if (key.includes('token')) {
            return '234lkj23l4kj234lkj234lkj234lkj23l4kj234l'
        }

        if (key.includes('_at')) {
            return '2017-10-10T16:00:00Z'
        }

        if (key.includes('_count')) {
            return 42
        }

        if (key.includes('id')) {
            return 1000 + id++
        }

        if (key.includes('node_id')) {
            return 'MDA6RW50aXR5MQ=='
        }

        if (key.includes('url')) {
            return value.replace(/[1-9][0-9]{2,10}/, '000000001')
        }

        return value
    }

    function decodeBuffer(fixture) {
        const response = isArray(fixture.response) ? fixture.response.join('') : fixture.response

        if (!isObject(response)) {
            try {
                // Decode the hex buffer that nock made
                const decoded = Buffer.from(response, 'hex')
                var unzipped = zlib.gunzipSync(decoded).toString('utf-8')
            } catch (err) {
                throw new Error(`Error decoding nock hex:\n${err}`)
            }
        }

        return JSON.parse(unzipped)
    }

    // This only executes when first recording the request, but not on subsequent requests
    function afterRecord(fixtures) {
        const normalizedFixtures = fixtures.map(fixture => {
            const isGzipped = fixture.rawHeaders.includes('gzip')
            let res = fixture.response

            if (fixture.body.note) {
                fixture.body.note = 'Hello from the inside!'
            }

            fixture.path = stripAccessToken(fixture.path)
            fixture.rawHeaders = fixture.rawHeaders.map(header => stripAccessToken(header))

            if (isGzipped) {
                res = decodeBuffer(fixture)
            }

            if (isArray(res)) {
                res = res.slice(0, 3).map(res => {
                    return mapValues(res, normalize)
                })
            } else {
                res = mapValues(res, normalize)
            }

            if (isGzipped) {
                try {
                    // Re-gzip to keep the octokittens happy
                    const stringified = JSON.stringify(res)
                    var zipped = zlib.gzipSync(stringified)
                } catch (err) {
                    throw new Error(`Error re-gzipping nock ==> ${err}`)
                }
            }

            fixture.response = zipped || res

            return fixture
        })

        return normalizedFixtures
    }

    function stripAccessToken(header) {
        return header.includes('access_token')
            ? header.replace(/access_token(.*?)(&|$)/gi, '')
            : header
    }

    function before(scope) {
        scope.filteringPath = () => stripAccessToken(scope.path)
        scope.filteringRequestBody = (body, aRecordedBody) => {
            if (body.includes('note')) {
                body = JSON.parse(body)
                body.note = aRecordedBody.note

                return JSON.stringify(body)
            }

            return body
        }
    }
}

function formatCmdName(cmd, argv) {
    if (argv.length === 1) {
        return cmd.name
    }

    return cmd.flags.reduce((flagName, current) => {
        if (flagName) {
            return flagName
        }

        if (argv.includes(current)) {
            return concatUpper(cmd.name, current.slice(2))
        }
    }, null)
}

function concatUpper(one, two) {
    return `${one}${upperFirst(two)}`
}
