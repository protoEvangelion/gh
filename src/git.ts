/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as exec from './exec'
import * as logger from './logger'

const git_command = process.env.GH_GIT_COMMAND || 'git'

const testing = process.env.NODE_ENV === 'testing'

export function checkout(branch, newBranch?: string) {
    var args = ['checkout', branch]

    if (newBranch) {
        args.push('-B', newBranch)
    }

    return !testing && exec.spawnSyncStream(git_command, args)
}

export function clone(url, folder) {
    var args = ['clone', url]

    if (folder) {
        args.push(folder)
    }

    return !testing && exec.spawnSyncStream(git_command, args)
}

export function _merge(branch, type) {
    try {
        const args = [type, branch]

        !testing && exec.spawnSyncStream(git_command, [...args])
    } catch (err) {
        if (err.code && err.code !== 0) {
            !testing && exec.spawnSyncStream(git_command, [type, '--abort'])
            throw err
        }
    }
}

export function merge(branch) {
    return !testing && this._merge(branch, 'merge')
}

export function rebase(branch) {
    return !testing && this._merge(branch, 'rebase')
}

export function push(remote, branch) {
    var args = ['push', remote]

    if (branch) {
        args.push(branch)
    }

    return !testing && exec.spawnSyncStream(git_command, args)
}

export function fetch(repoUrl, headBranch, pullBranch) {
    var args = ['fetch', repoUrl, `${headBranch}:${pullBranch}`, '--no-tags']

    return !testing && exec.spawnSyncStream(git_command, args)
}

export function countUserAdjacentCommits() {
    let git
    let params
    let commits = 0
    const user = getConfig('user.name')
    let author

    do {
        params = ['log', '-1', `--skip=${commits}`, '--pretty=%an']
        git = exec.spawnSync(git_command, params)

        if (git.status !== 0) {
            logger.error(git.stderr)
        }

        author = git.stdout

        commits += 1
    } while (author === user)

    commits -= 1

    return commits
}

export function deleteBranch(branch) {
    if (testing) {
        return
    }

    var git = exec.spawnSync(git_command, ['branch', '-d', branch])

    if (git.status !== 0) {
        logger.debug(git.stderr)
    }

    return git.stdout
}

export function findRoot() {
    return exec.spawnSync(git_command, ['rev-parse', '--show-toplevel']).stdout
}

export function getCommitMessage(branch, number) {
    let git
    const params = ['log']

    if (!number) {
        number = 1
    }

    params.push(`-${number}`, '--first-parent', '--no-merges', '--pretty=%s')

    if (branch) {
        params.push(branch)
    }

    params.push('--')

    git = exec.spawnSync(git_command, params)

    if (git.status !== 0) {
        logger.debug("Can't get commit message.")
        return
    }

    return git.stdout
}

export function getConfig(key) {
    var git = exec.spawnSync(git_command, ['config', '--get', key])

    if (git.status !== 0) {
        throw new Error(`No git config found for ${key}\n`)
    }

    return git.stdout
}

export function getCurrentBranch() {
    if (testing) {
        return 'master'
    }

    var git = exec.spawnSync(git_command, ['symbolic-ref', '--short', 'HEAD'])

    if (git.status !== 0) {
        logger.debug("Can't get current branch.")
        return
    }

    return git.stdout
}

export function getLastCommitMessage(branch) {
    return getCommitMessage(branch, 1)
}

export function getLastCommitSHA() {
    var git = exec.spawnSync(git_command, ['rev-parse', '--short', 'HEAD'])

    if (git.status !== 0) {
        throw new Error("Can't retrieve last commit.")
    }

    return git.stdout
}

export function getRemoteUrl(remote) {
    try {
        return getConfig(`remote.${remote}.url`)
    } catch (e) {
        logger.debug("Can't get remote URL.")
        return
    }
}

export function getRepoFromRemoteURL(url) {
    var parsed = parseRemoteUrl(url)

    return parsed && parsed[1]
}

export function getUserFromRemoteUrl(url) {
    var parsed = parseRemoteUrl(url)

    return parsed && parsed[0]
}

export function getRepo(remote) {
    return getRepoFromRemoteURL(getRemoteUrl(remote))
}

export function getUser(remote) {
    return getUserFromRemoteUrl(getRemoteUrl(remote))
}

export function parseRemoteUrl(url) {
    var parsed = /[\/:]([\w-]+)\/(.*?)(?:\.git)?$/.exec(url)

    if (parsed) {
        parsed.shift()
    }

    return parsed
}
