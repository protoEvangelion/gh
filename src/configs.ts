/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as fs from 'fs'
import { cloneDeep } from 'lodash'
import * as path from 'path'
import * as R from 'ramda'
import * as S from 'sanctuary'
import * as userhome from 'userhome'
import * as which from 'which'
import * as exec from './exec'
import * as logger from './logger'
import { safeImport } from './fp'

export const PLUGINS_PATH_KEY = 'plugins_path'

const testing = process.env.NODE_ENV === 'testing'

/* Refactored FP Functions */

const safeWhich = S.encase(which.sync)
const safeRealpath = S.encase(fs.realpathSync)

export const getPluginPath = R.pipeK(
    safeWhich,
    safeRealpath
)

// TODO merge with getPlugin fn in cmd.ts
export const getPlugin = R.pipeK(
    S.prepend('gh-'),
    getPluginPath,
    safeImport
)

/* ----------------------- */

// function concatError(leftMonad, errorMessage) {
//     const leftMonadContainsError = leftMonad && leftMonad.isLeft() && leftMonad.value

//     return leftMonadContainsError ? leftMonad.map(mapError) : Left(errorMessage)

//     function mapError(prevErrorMessage) {
//         return `${prevErrorMessage}\n${errorMessage}`
//     }
// }

// -- Config -------------------------------------------------------------------

export function getNodeModulesGlobalPath() {
    try {
        var { stdout } = exec.spawnSync('npm', ['root', '-g'])
    } catch (err) {
        logger.warn(`Can't resolve plugins directory path.\n${err}`)
    }

    return stdout
}

export function getGlobalPackageJson() {
    const configFile = fs.readFileSync(path.join(__dirname, '../package.json'))

    return JSON.parse(configFile.toString())
}

export function getDefaultConfigPath() {
    return path.join(__dirname, '../default.gh.json')
}

export function getProjectConfigPath() {
    return path.join(process.cwd(), '.gh.json')
}

export function getUserHomePath() {
    return userhome('.gh.json')
}

function resolveGHConfigs() {
    const globalConfig = getGlobalConfig()
    let projectConfig
    const result = {}

    try {
        projectConfig = JSON.parse(fs.readFileSync(getProjectConfigPath()).toString())

        Object.keys(globalConfig).forEach(key => {
            result[key] = globalConfig[key]
        })

        Object.keys(projectConfig).forEach(key => {
            result[key] = projectConfig[key]
        })

        return result
    } catch (e) {
        logger.debug(e.message)

        if (e.code !== 'MODULE_NOT_FOUND' && e.code !== 'ENOENT') {
            throw e
        }

        return globalConfig
    }
}

export function getConfig() {
    const config = resolveGHConfigs()

    const protocol = `${config.api.protocol}://`
    const is_enterprise = config.api.host !== 'api.github.com'

    if (config.github_host === undefined) {
        config.github_host = `${protocol}${is_enterprise ? config.api.host : 'github.com'}`
    }
    if (config.github_gist_host === undefined) {
        config.github_gist_host = `${protocol}${
            is_enterprise ? `${config.api.host}/gist` : 'gist.github.com'
        }/`
    }

    return config
}

export function getGlobalConfig() {
    const configPath = getUserHomePath()
    const defaultPath = getDefaultConfigPath()

    if (!fs.existsSync(configPath)) {
        createGlobalConfig()
    }

    return JSON.parse(fs.readFileSync(testing ? defaultPath : configPath).toString())
}

export function removeGlobalConfig(key) {
    var config = getGlobalConfig()

    delete config[key]

    saveJsonConfig(getUserHomePath(), config)
}

export function createGlobalConfig() {
    saveJsonConfig(
        getUserHomePath(),
        JSON.parse(fs.readFileSync(getDefaultConfigPath()).toString())
    )
}

export function writeGlobalConfig(jsonPath, value) {
    const config = getGlobalConfig()
    let i
    let output
    let path
    let pathLen

    path = jsonPath.split('.')
    output = config

    for (i = 0, pathLen = path.length; i < pathLen; i++) {
        output[path[i]] = config[path[i]] || (i + 1 === pathLen ? value : {})
        output = output[path[i]]
    }

    saveJsonConfig(getUserHomePath(), config)
}

export function saveJsonConfig(path, object) {
    fs.writeFileSync(path, JSON.stringify(object, null, 4))
}

export function writeGlobalConfigCredentials(user, token, path): void {
    const configPath = path || getUserHomePath()

    let config
    if (fs.existsSync(configPath)) {
        config = JSON.parse(fs.readFileSync(configPath).toString())
    } else {
        config = JSON.parse(fs.readFileSync(getDefaultConfigPath()).toString())
    }

    logger.log(`Writing GH config data: ${configPath}`)

    try {
        config.github_user = user
        config.github_token = token

        saveJsonConfig(configPath, config)
    } catch (err) {
        throw new Error(`Error writing credentials to global config\n${err}`)
    }

    logger.log('Authentication succeed. Token written to global config.')
}

// -- Plugins ------------------------------------------------------------------

export function addPluginConfig(plugin) {
    try {
        const pluginConfig = require(path.join(
            getNodeModulesGlobalPath(),
            `gh-${plugin}`,
            'gh-plugin.json'
        ))

        const config = getGlobalConfig()
        const configHooks = cloneDeep(config.hooks)
        const pluginHooks = cloneDeep(pluginConfig.hooks)

        if (config.plugins[plugin] && !config.plugins[plugin]['hooks_installed']) {
            Object.keys(pluginHooks).forEach(cmd => {
                Object.keys(pluginHooks[cmd]).forEach(hook => {
                    configHooks[cmd][hook].before = [
                        ...configHooks[cmd][hook].before,
                        ...pluginHooks[cmd][hook].before,
                    ]

                    configHooks[cmd][hook].after = [
                        ...configHooks[cmd][hook].after,
                        ...pluginHooks[cmd][hook].after,
                    ]
                })
            })

            if (!testing) {
                logger.log(
                    logger.colors.yellow(
                        `Copying over ${plugin} plugin hooks to your .gh.json hooks.`
                    )
                )

                try {
                    config.hooks = configHooks
                    config.plugins[plugin]['hooks_installed'] = true

                    saveJsonConfig(getUserHomePath(), config)
                } catch (err) {
                    logger.error(`Error writing ${plugin} hooks to .gh.json config.\n${err}`)
                }

                logger.log(logger.colors.green('Copy successful.\n'))
            }
        }
    } catch (e) {
        if (e.code !== 'MODULE_NOT_FOUND') {
            throw e
        }
    }
}

export function getPlugins() {
    const pluginsPath = getNodeModulesGlobalPath()

    if (pluginsPath === '') {
        return []
    }

    try {
        var plugins = fs.readdirSync(pluginsPath).filter(plugin => {
            return plugin.substring(0, 3) === 'gh-'
        })
    } catch (err) {
        logger.warn(`Can't read plugins directory.\n${err}`)
    }

    return plugins
}

export function pluginHasConfig(pluginName) {
    return Boolean(getConfig().plugins[pluginName])
}
