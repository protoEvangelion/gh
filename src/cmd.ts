/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

// -- Requires -------------------------------------------------------------------------------------

import * as fs from 'fs'
import * as nopt from 'nopt'
import * as path from 'path'
import * as updateNotifier from 'update-notifier'
import { find, getUser } from './base'
import * as configs from './configs'
import * as git from './git'
import { produce, setAutoFreeze } from 'immer'

const config = configs.getConfig()

// allows to run program as js or ts
const extension = __filename.slice(__filename.lastIndexOf('.') + 1)

const testing = process.env.NODE_ENV === 'testing'

// -- Utils ----------------------------------------------------------------------------------------

async function resolveCmd(name, commandDir) {
    const reg = new RegExp(`.${extension}$`, 'i')
    const commandFiles = find(commandDir, reg)

    const commandName = commandFiles.filter(file => {
        switch (file) {
            case `milestone.${extension}`:
                if (name === 'ms') return true
                break
            case `notification.${extension}`:
                if (name === 'nt') return true
                break
            case `pull-request.${extension}`:
                if (name === 'pr') return true
                break
        }

        if (file.startsWith(name)) {
            return true
        }

        return false
    })[0]

    return commandName && import(path.join(commandDir, commandName))
}

async function resolvePlugin(name) {
    // If plugin command exists, register the executed plugin name
    process.env.NODEGH_PLUGIN = name

    const plugin = await configs.getPlugin(name)
    const pluginFullName = plugin.Impl.name.toLowerCase()

    plugin && configs.addPluginConfig(pluginFullName)

    return plugin
}

async function loadCommand(name) {
    let Command

    const commandDir = path.join(__dirname, 'cmds')
    const commandPath = path.join(commandDir, `${name}.${extension}`)

    if (fs.existsSync(commandPath)) {
        Command = await import(commandPath)
    } else {
        Command = await resolveCmd(name, commandDir)
    }

    if (!Command) {
        // try to resolve as plugin
        const plugin = await resolvePlugin(name)

        Command = { default: plugin.Impl }
    }

    return Command.default
}

function notifyVersion() {
    var notifier = updateNotifier({ pkg: configs.getGlobalPackageJson() })

    if (notifier.update) {
        notifier.notify()
    }
}

async function getCommand(args) {
    /**
     * nopt function returns:
     *
     * remain: The remaining args after all the parsing has occurred.
     * original: The args as they originally appeared.
     * cooked: The args after flags and shorthands are expanded.
     */
    const parsed = nopt(args)
    const remain = parsed.argv.remain
    const cooked = parsed.argv.cooked
    let module = remain[0]

    if (cooked[0] === '--version' || cooked[0] === '-v') {
        module = 'version'
    } else if (!remain.length || cooked.indexOf('-h') >= 0 || cooked.indexOf('--help') >= 0) {
        module = 'help'
    }

    try {
        var Command = await loadCommand(module)
    } catch (err) {
        throw new Error(`Cannot find module ${module}\n${err}`)
    }

    if (!Command) {
        throw new Error(`No cmd or plugin found.`)
    }

    return Command
}

export async function setUp() {
    notifyVersion()

    const Command = await getCommand(process.argv)

    /**
     * If you run `gh pr 1 -s node-gh --remote=origin --user protoEvangelion`, nopt will return
     *
     *   {
     *     remote: 'origin',
     *     submit: 'node-gh',
     *     user: 'protoEvangelion',
     *     argv: {
     *         original: ['pr', '1', '-s', 'pr', 'node-gh', '--remote', 'origin', '--user', 'protoEvangelion'],
     *         remain: ['pr', '1'],
     *         cooked: ['pr', '1', '--submit', 'node-gh', '--remote', 'origin', '--user', 'protoEvangelion'],
     *     },
     *   }
     *
     * Historically we passed every arg after 2nd arg (gh pr 1 -s user; everything after 'pr')
     * and all parsed options to each cmd's payload function to figure out positional args and allow for neat shortcuts like:
     * gh is 'new issue' 'new issue description'
     */
    const args = nopt(Command.DETAILS.options, Command.DETAILS.shorthands, process.argv, 2)

    setAutoFreeze(false)

    const options = produce(args, draft => {
        // Gets 2nd positional arg (`gh pr 1` will return 1)
        const secondArg = [draft.argv.remain[1]]
        const remote = draft.remote || config.default_remote
        const remoteUrl = git.getRemoteUrl(remote)

        draft.remote = remote
        draft.number = draft.number || secondArg
        draft.loggedUser = getUser()
        draft.remoteUser = git.getUserFromRemoteUrl(remoteUrl)
        draft.repo = draft.repo || git.getRepoFromRemoteURL(remoteUrl)
        draft.currentBranch = git.getCurrentBranch()
        draft.github_host = config.github_host
        draft.github_gist_host = config.github_gist_host

        if (!draft.user) {
            if (draft.repo || draft.all) {
                draft.user = draft.loggedUser
            } else {
                draft.user = process.env.GH_USER || draft.remoteUser || draft.loggedUser
            }
        }

        /**
         * Checks if there are aliases in your .gh.json file.
         * If there are aliases in your .gh.json file, we will attempt to resolve the user, PR forwarder or PR submitter to your alias.
         */
        if (config.alias) {
            draft.fwd = config.alias[draft.fwd] || draft.fwd
            draft.submit = config.alias[draft.submit] || draft.submit
            draft.user = config.alias[draft.user] || draft.user
        }
    })

    if (testing) {
        const { prepareTestFixtures } = await import('./utils')

        await new Command().run(options, prepareTestFixtures(Command.name, args.argv.cooked))
    } else {
        await new Command().run(options)
    }
}

export async function run() {
    if (!fs.existsSync(configs.getUserHomePath())) {
        configs.createGlobalConfig()
    }

    try {
        process.env.GH_PATH = path.join(__dirname, '../')

        await setUp()
    } catch (e) {
        console.error(e.stack || e)
    }
}
