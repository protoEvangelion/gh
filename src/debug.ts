#!/usr/bin/env node

/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import { run } from './cmd'

const verbose = process.argv.indexOf('--verbose') !== -1
const insane = process.argv.indexOf('--insane') !== -1

process.on('unhandledRejection', r => console.log(r))

if (verbose || insane) {
    process.env.GH_VERBOSE = 'true'
}

if (insane) {
    process.env.GH_VERBOSE_INSANE = 'true'
}

run()
