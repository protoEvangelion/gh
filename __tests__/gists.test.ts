/**
 * Â© 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as stripAnsi from 'strip-ansi'
import { runCmd } from '../src/test-utils'

describe('E2E: Gist Module Test', () => {
    it('List Gists `gh gi`', done => {
        expect(runCmd('gh gi')).toMatchSnapshot()
        done()
    })

    it('Open gist in browser `gh gi --browser --id 5444883`', done => {
        expect(runCmd('gh gi --browser --id 5444883')).toMatchSnapshot()
        done()
    })

    it('Create new gist `gh gi --new hello --content "Hello World!"`', done => {
        expect(runCmd('gh gi --new hello --content "Hello World!"')).toMatchSnapshot()
        done()
    })

    it('Fork a gist `gh gi --fork 5444883`', done => {
        expect(runCmd('gh gi --fork 5444883')).toMatchSnapshot()
        done()
    })

    it('Delete gists `gh gi --delete 5250d21093b46bd0665c2e8656d16bd2 --delete 30e5d3c69a6997617ab69d07b733105e`', done => {
        expect(
            stripAnsi(
                runCmd(
                    'echo "y\n" | gh gi --delete 5250d21093b46bd0665c2e8656d16bd2 --delete 30e5d3c69a6997617ab69d07b733105e'
                )
            )
        ).toMatchSnapshot()
        done()
    })
})
