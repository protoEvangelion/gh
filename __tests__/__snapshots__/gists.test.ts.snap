// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E: Gist Module Test Create new gist \`gh gi --new hello --content "Hello World!"\` 1`] = `
"Creating public gist on protoEvangelion
https://gist.github.com/000000001e9f6422a8d3cd0feda0a3be845d4
[hook] gh gi --browser  --id 1000
"
`;

exports[`E2E: Gist Module Test Delete gists \`gh gi --delete 5250d21093b46bd0665c2e8656d16bd2 --delete 30e5d3c69a6997617ab69d07b733105e\` 1`] = `
"? Are you sure? This action CANNOT be undone. [y/N] ? Are you sure? This action CANNOT be undone. [y/N] y? Are you sure? This action CANNOT be undone. [y/N] y
Deleting gist protoEvangelion/5250d21093b46bd0665c2e8656d16bd2
✓ Successfully deleted gist: 5250d21093b46bd0665c2e8656d16bd2
Deleting gist protoEvangelion/30e5d3c69a6997617ab69d07b733105e
✓ Successfully deleted gist: 30e5d3c69a6997617ab69d07b733105e
"
`;

exports[`E2E: Gist Module Test Fork a gist \`gh gi --fork 5444883\` 1`] = `
"Forking gist on protoEvangelion
https://gist.github.com/c93ab6d93dde5dfa3fe9f0b000000001b1643c
[hook] gh gi --browser  --id
"
`;

exports[`E2E: Gist Module Test List Gists \`gh gi\` 1`] = `
"Listing gists for protoEvangelion
protoEvangelion/1000 in 4 years
Hubspot Form Generator
https://gist.github.com/000000001f227a2c690e76c6fa2ca5b444de1

protoEvangelion/1004 in 4 years
Liferay.provide minimal example
https://gist.github.com/000000001b5c7f25c00095ef73095db0b59cfc

protoEvangelion/1008 in 4 years
Fix aui adding pre styles
https://gist.github.com/dfc000000001b8c49c8a97dbc98cf7193

"
`;

exports[`E2E: Gist Module Test Open gist in browser \`gh gi --browser --id 5444883\` 1`] = `
"https://gist.github.com/5444883
"
`;
